c/parallel/src/radix_sort.cu

```diff
@@ -0,0 +1,843 @@
+    build_ptr->cubin_size = result.size;
+    build_ptr->key_type   = input_keys_it.value_type;
+    build_ptr->value_type = input_values_it.value_type;
+    build_ptr->order      = sort_order;
+  }
+  catch (const std::exception& exc)
+  {
+    fflush(stderr);
+    printf("\nEXCEPTION in cccl_device_radix_sort_build(): %s\n", exc.what());
+    fflush(stdout);
+    error = CUDA_ERROR_UNKNOWN;
+  }
+
+  return error;
+}
+
+template <cub::SortOrder Order>
+CUresult cccl_device_radix_sort(
+  cccl_device_radix_sort_build_result_t build,
+  void* d_temp_storage,
+  size_t* temp_storage_bytes,
+  cccl_iterator_t d_keys_in,
+  cccl_iterator_t d_keys_out,
+  cccl_iterator_t d_values_in,
+  cccl_iterator_t d_values_out,
+  cccl_op_t decomposer,
+  uint64_t num_items,
+  int begin_bit,
+  int end_bit,
+  bool is_overwrite_okay,
+  int* selector,
+  CUstream stream)
+{
+  if (cccl_iterator_kind_t::CCCL_POINTER != d_keys_in.type || cccl_iterator_kind_t::CCCL_POINTER != d_values_in.type
+      || cccl_iterator_kind_t::CCCL_POINTER != d_keys_out.type
+      || cccl_iterator_kind_t::CCCL_POINTER != d_values_out.type)
+  {
+    fflush(stderr);
+    printf("\nERROR in cccl_device_radix_sort(): radix sort input must be a pointer\n");
+    fflush(stdout);
+    return CUDA_ERROR_UNKNOWN;
+  }
+
+  CUresult error = CUDA_SUCCESS;
+  bool pushed    = false;
+  try
+  {
+    pushed = try_push_context();
+
+    CUdevice cu_device;
+    check(cuCtxGetDevice(&cu_device));
+
+    cub::DoubleBuffer<indirect_arg_t> d_keys_buffer(
+      static_cast<indirect_arg_t*>(d_keys_in.state), static_cast<indirect_arg_t*>(d_keys_out.state));
+    cub::DoubleBuffer<indirect_arg_t> d_values_buffer(
+      static_cast<indirect_arg_t*>(d_values_in.state), static_cast<indirect_arg_t*>(d_values_out.state));
+
+    cub::DispatchRadixSort<Order,
+                           indirect_arg_t,
+                           indirect_arg_t,
+                           OffsetT,
+                           indirect_arg_t,
+                           radix_sort::dynamic_radix_sort_policy_t<&radix_sort::get_policy>,
+                           radix_sort::radix_sort_kernel_source,
+                           cub::detail::CudaDriverLauncherFactory>::
+      Dispatch(
+        d_temp_storage,
+        *temp_storage_bytes,
+        d_keys_buffer,
+        d_values_buffer,
+        num_items,
+        begin_bit,
+        end_bit,
+        is_overwrite_okay,
+        stream,
+        decomposer,
+        {build},
+        cub::detail::CudaDriverLauncherFactory{cu_device, build.cc},
+        {d_keys_in.value_type.size});
+
+    *selector = d_keys_buffer.selector;
+  }
+  catch (const std::exception& exc)
+  {
+    fflush(stderr);
+    printf("\nEXCEPTION in cccl_device_radix_sort(): %s\n", exc.what());
+    fflush(stdout);
+    error = CUDA_ERROR_UNKNOWN;
+  }
+
+  if (pushed)
+  {
+    CUcontext dummy;
+    cuCtxPopCurrent(&dummy);
+  }
+
+  return error;
+}
+
+CUresult cccl_device_ascending_radix_sort(
```

Comment:
Why not have single `cccl_device_radix_sort` which dispatches at run-time to `cccl_device_radix_sort<cub::SortOrder::Ascending>` or `cccl_device_radix_sort<cub::SortOrder::Descending>`?